core/VectorState.hpp

```diff
@@ -34,85 +80,411 @@ namespace atlas
 
         inline void setVSTART(uint8_t value) { vstart_ = value; }
 
-        template <typename VLEN> inline uint8_t getVLMAX()
+        // helper functions
+
+        inline uint8_t getVLMAX() const { return getVLEN() / 8 * getLMUL() / getSEW(); }
+
+        template <typename XLEN> XLEN vsetAVL(AtlasState* state, bool set_max, XLEN avl = 0)
         {
-            return sizeof(VLEN) * lmul_ / (sew_ == 0 ? 1 : sew_);
+            uint8_t vl = set_max ? getVLMAX() : std::min<uint8_t>(getVLMAX(), avl);
```

Comment:
Variables that are assigned during creating and not assigned again should be `const`